Known Customer Credential-Issuer (#1330)
TBD54566975/developer.tbd.website: site/docs/tbdex/issuer/kcc-issuer.mdx

Reviewer: Need to put this in a new folder called Known Customer Credential

![image](https://github.com/TBD54566975/developer.tbd.website/assets/15972783/f299d36f-3f2a-4ba8-a252-27f0ae18d5fb)

---
title: Issuer Flow
hide_title: true
---

> <LanguageSwitcher languages="JavaScript, Kotlin, Swift" />

Reviewer: ```suggestion
<LanguageSwitcher languages="JavaScript, Kotlin" />
```

Swift not supported

# Known Customer Credential

> Known Customer Credentials (KCCs) are [Verifiable Credentials](https://developer.tbd.website/docs/web5/learn/verifiable-credentials) designed to streamline the [Know Your Customer(KYC)](/docs/glossary) process for tbDEX protocol users. KCCs help in gaining access to PFIs providing regulated financial services. For a comprehensive exploration of the practical applications and compliance considerations of KCCs, refer to our [KCC Guide](https://docs.google.com/document/d/1vwcqzXOpVrCUv7U7g3A4jJs99Lj18gAVy4FaA3COEb8/edit). This resource provides detailed examples and insights into how KCCs align with certain regulatory requirements, offering valuable context for those overseeing the implementation and management of KCCs.

Reviewer: ```suggestion
Known Customer Credentials (KCCs) are [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) designed to streamline the Know Your Customer(KYC) process for tbDEX protocol users. KCCs help in gaining access to PFIs providing regulated financial services. For a comprehensive exploration of the practical applications and compliance considerations of KCCs, refer to our [KCC Guide](/). This resource provides detailed examples and insights into how KCCs align with certain regulatory requirements, offering valuable context for those overseeing the implementation and management of KCCs.
```

Reviewer: please open an Asana issue to make sure we come back and update this with the correct link for the KCC Guide once it's published

> :::danger

Reviewer: ```suggestion
:::danger Important
```

this changes the callout box title to say "IMPORTANT" instead of "DANGER"

> IMPORTANT: This guide is intended for educational purposes only and does not constitute legal advice. Compliance programs may have varying requirements. Consult your legal and/or compliance advisors to ensure that the KCC is consistent with your legal and compliance obligations.

Reviewer: ```suggestion
This guide is intended for educational purposes only and does not constitute legal advice. Compliance programs may have varying requirements. Consult your legal and/or compliance advisors to ensure that the KCC is consistent with your legal and compliance obligations.
```

:::

In this guide, we'll cover:

> - Performing [Identity Verification (IDV)](https://csrc.nist.gov/glossary/term/identity_verification#:~:text=1%20under%20Verification-,The%20process%20of%20confirming%20or%20denying%20that%20a%20claimed%20identity,with%20the%20identity%20being%20claimed) as a necessary step **before** issuing a Known Customer Credential (KCC), following [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html) set of specifications.

Reviewer: ```suggestion
- Performing Identity Verification (IDV) as a necessary step **before** issuing a Known Customer Credential (KCC), following [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) and [OpenID for Verifiable Presentation's (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html) set of specifications.
```

- Designing a KCC, detailing the required properties.
- Issuing a KCC, following successful IDV.

:::info

> For a detailed guide on Known Customer Credentials for Wallets, please refer to our [Wallet's Known Customer Credential Guide](https://hackmd.io/SQwE3SnrRkmeipGGtznJTw)

Reviewer: ```suggestion
For a detailed guide on managing Known Customer Credentials as a Wallet application, please refer to the [Known Customer Credential Guide for Wallet Apps](/)
```

make Asana issue to update link once published

:::

## Environment Setup

> If you haven't already, please follow our [Credential Issuance Server Setup Guide](https://hackmd.io/4cW3D9BVQGCEuf1CEg47jA?both#Credential-Issuance-Enviornment-Setup-Guide) for detailed instructions on the dependencies and packages needed to set up your server.

Reviewer: ```suggestion
If you haven't already, please follow the [Credential Issuance Server Setup Guide](/) for detailed instructions on the dependencies and packages needed to set up your server.
```

Create Asana issue to update link once published

## IDV Process

Identity Verification (IDV) is a critical component of typical Know Your Customer (KYC) requirements, where certain types of Personally Identifying Information (PII) are collected and verified from an individual. IDV is a crucial step before a Known Customer Credential (KCC) can be issued.

> [**IDV Flow diagram**]

Reviewer: let's create an Asana ticket to track this

> The IDV flow begins when the customer's agent (e.g. Wallet application) sends an HTTP request to the Issuer's [IDV service endpoint](https://hackmd.io/4cW3D9BVQGCEuf1CEg47jA?both#Issuers-Identity) specified in the Issuer's [Decentralized Identifier (DID) Document](/docs/glossary#did-document).

Reviewer: ```suggestion
The IDV flow begins when the customer's agent (e.g. Wallet application) sends an HTTP request to the Issuer's [IDV service endpoint](https://hackmd.io/4cW3D9BVQGCEuf1CEg47jA?both#Issuers-Identity) specified in the Issuer's [Decentralized Identifier (DID) Document](/docs/glossary#did-document).
```

Create Asana ticket (one big one is fine) to track this link change

<Shnip
  snippets={[
    {
      snippetName: 'siopv2InitiateId_tokenEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js for id_token only ',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

:::note
`id_token` is a Self Issued Identity Token, that authenticates the customer to the Issuer by sharing the customer's Decentralized Identifier. This guide will cover an example `id_token` to learn more refer to [openID's specs](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-id-token).
:::

### Approach 2: Requesting both `id_token` and [`vp_token`](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2)

This approach is used when you need to authenticate the user and also want to request specific Verifiable Credentials:

<Shnip
  snippets={[
    {
      snippetName: 'siopv2InitiateId_tokenAndVp_tokenEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js for id_token and vp_token',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

> The SIOPv2 Authorization request contains information about you, the issuer, along with a [Presentation Definition](/docs/web5/build/verifiable-credentials/presentation-definition).

Reviewer: ```suggestion
The SIOPv2 Authorization request contains information about you, the Issuer, along with a [Presentation Definition](/docs/web5/build/verifiable-credentials/presentation-definition).
```

- **`client_id`**: Issuer's DID, establishing the issuer/requestor's identity.

> - **`response_type`**: Lets the wallet know the desired format, with `id_token` being required identity token and `vp_token` for _optional_ verifiable presentation token. The inclusion of `vp_token` means a presentation definition is required within the request.

Reviewer: ```suggestion
- **`response_type`**: Lets the Wallet know the desired format, with `id_token` being required identity token and `vp_token` for _optional_ verifiable presentation token. The inclusion of `vp_token` means a presentation definition is required within the request.
```

> - **`response_uri`**: Indicates the callback url for the wallet, ensuring direct communication between the wallet and the issuer.

Reviewer: ```suggestion
- **`response_uri`**: Indicates the callback url for the Wallet, ensuring direct communication between the Wallet and the Issuer.
```

- **`nonce`**: A unique identifier tying together the request and its response.

:::note

> `vp_token` is a [Verifiable Presentation](/docs/glossary#verifiable-presentation) that contains certified and authenticated claims about the customer. In this guide we'll cover an example `vp_token` to learn more refer to [openID's specs](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2).

Reviewer: ```suggestion
`vp_token` is a [Verifiable Presentation](/docs/glossary#verifiable-presentation) that contains certified and authenticated claims about the customer. In this guide, we'll cover an example `vp_token`. To learn more refer to [OpenID's specs](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2).
```

:::

2. Encode the SIOPv2 Authorization Request and respond:

<Shnip
  snippets={[
    {
      snippetName: 'encodeSiopv2AuthRequestIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

3. The Wallet will then respond with a POST SIOPv2 Authorization Response. This response will be [JSON Web Tokens (JWTs)](https://jwt.io/). Here's an example response where `vp_token` and `id_token` is in the Authorization Response:

```javascript

> // Wallet id_token JWT response

Reviewer: i think you wanted this ('Wallet.js') to be the name of the file or no?

```

```javascript
// Wallet vp_token JWT response

eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9
  .eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVQcmVzZW50YXRpb24iXSwidmVyaWZpYWJsZUNyZWRlbnRpYWwiOlt7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIiwiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvZXhhbXBsZXMvdjEiXSwiaWQiOiJodHRwczovL2V4YW1wbGUuY29tL2NyZWRlbnRpYWxzLzE4NzIiLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSURDYXJkQ3JlZGVudGlhbCJdLCJpc3N1ZXIiOnsiaWQiOiJkaWQ6ZXhhbXBsZTppc3N1ZXIifSwiaXNzdWFuY2VEYXRlIjoiMjAxMC0wMS0wMVQxOToyMzoyNFoiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJnaXZlbl9uYW1lIjoiRnJlZHJpayIsImZhbWlseV9uYW1lIjoiU3RyJm9tYmVyZyIs
  .ImJpcnRoZGF0ZSI6IjE5NDktMDEtMjIifX1dLCJob2xkZXIiOiJkaWQ6ZXhhbXBsZTpob2xkZXIiLCJwcm9vZiI6eyJ0eXBlIjoiRWQyNTUxOVNpZ25hdHVyZTIwMTgiLCJjcmVhdGVkIjoiMjAyMS0wMy0xOVQxNTozMDoxNVoiLCJjaGFsbGVuZ2UiOiJuLTBTNl9XekEyTWoiLCJkb21haW4iOiJodHRwczovL2NsaWVudC5leGFtcGxlLm9yZy9jYiIsImp3cyI6ImV5SmhiR2NpT2lKSVV6STFOaUo5LkhvbGRlcl9GT1JfU0lHTkFUVVJFIiwicHJvb2ZQdXJwb3NlIjoiYXV0aGVudGljYXRpb24iLCJ2ZXJpZmljYXRpb25NZXRob2QiOiJkaWQ6ZXhhbXBsZTpob2xkZXIja2V5LTEifX0;
```

> When decoded here is the Human-Readable JSON of `id_token` and `vp_token`:

Reviewer: ```suggestion
When decoded, here is the JSON of `id_token` and `vp_token`:
```

    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

:::info
When the Wallet's response solely contains `id_token`, your response must include a **`url`** (IDV Request Form). This will point to a webpage for the customer to input their information (e.g.; name, identification number, date of birth, etc).  
:::

#### Credential Offer Details:

- **`credential_issuer`**: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's metadata and the Issuer's Authorization Server Metadata.
- **`credential_configuration_ids`**: Unique strings that identify the credential you are offering. The Wallet can use these to request metadata.
- **`grants`**: Contains grant types you will accept for this credential offer.
- **`urn:ietf:params:oauth:grant-type:pre-authorized_code`**: Grant type that allows the Wallet to follow a Pre-Authorized code flow skipping any PII steps because they've already completed it.
- **`pre-authorized_code`**: The code representing the Credential Issuer's authorization for the Wallet to obtain an access token.

:::info
**Dynamic Metadata Retrieval by Wallets**

> Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching issuer and authorization server metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and` /.well-known/oauth-authorization-server` to the `credential_issuer`, allow wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.

Reviewer: ```suggestion
Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching issuer and authorization server metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and` /.well-known/oauth-authorization-server` to the `credential_issuer`, allow Wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.
```

> Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching issuer and authorization server metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and` /.well-known/oauth-authorization-server` to the `credential_issuer`, allow wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.

Reviewer: ```suggestion
Wallets use the `credential_issuer` from the Credential Offer to dynamically construct URLs for fetching Issuer and Authorization Server Metadata. These URLs, formed by appending `/.well-known/openid-credential-issuer` and` /.well-known/oauth-authorization-server` to the `credential_issuer`, allow wallets to access up-to-date information on issuer capabilities and OAuth 2.0 endpoints. The next steps will cover how to properly configure these endpoints.
```

:::

5. Configure the `/.well-known/openid-credential-issuer` endpoint that will return the Issuer's Metadata:

<Shnip
  snippets={[
    {
      snippetName: 'wellKnownIssuerMetadataEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

- `credential_issuer`: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's metadata and the Issuer's Authorization Server Metadata.

> - `credential_endpoint`: The endpoint the wallet will use to submit requests to issue credentials. After obtaining an access token, the wallet sends a credential request to this endpoint.

Reviewer: ```suggestion
- `credential_endpoint`: The endpoint the Wallet will use to submit requests to issue credentials. After obtaining an access token, the Wallet sends a credential request to this endpoint.
```

- `credential_configurations_supported`: Defines the format `jwt_vc_json` and the credential.

6. Configure the `/.well-known/oauth-authorization-server` endpoint that will return the Issuer's Authorization Server Metadata:

<Shnip
  snippets={[
    {
      snippetName: 'wellKnownAuthorizationServerEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

> - `issuer`: The base URL of the issuer's service.

Reviewer: ```suggestion
- `issuer`: The base URL of the Issuer's service.
```

- `token_endpoint`: The endpoint where Wallet's can exchange the `pre_authorization_code` receieved in the `credential_offer` for an access token.

7. Set up the `/token` endpoint to process incoming requests from the Wallet containing the `pre_authorization_code`. This code, originally provided in the `credential_offer` object, is used to validate the Wallet's request:

<Shnip
  snippets={[
    {
      snippetName: 'accessTokenEndpointIssuerJs',
      language: 'JavaScript',
      title: 'Issuer api.js',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

## Design Known Customer Credential

> When creating a Verifiable Credential you can design a model class to represent the specific type of credential you'd like to issue.

Reviewer: ```suggestion
When creating a Verifiable Credential, you can design a model class to represent the specific type of credential you'd like to issue.
```

1. Define a `KccCredential` class with the required fields for the credential:

<Shnip
  snippets={[
    {
      snippetName: 'KnownCustomerCredentialsClass',
      language: 'JavaScript',
      title: 'Issuer kccCredential.js',
    },
    {
      snippetName: '',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

:::info

- You'll need to assign `issuer`, `subject`, and `expirationDate` properties during the creation process. `issuanceDate` is managed and generated by the `Web5` SDK. You'll see how this works in the upcoming section.  
  :::

## Issue Known Customer Credential

> Issuing a Known Customer Credential to a customer's Wallet confirms the successful completion of the Identity Verification Process. To create and securely sign this credential we will utilize the [web5/credentials](https://www.npmjs.com/package/@web5/credentials) package.

Reviewer: ```suggestion
Issuing a Known Customer Credential to a customer's Wallet confirms the successful completion of the Identity Verification Process. To create and securely sign this credential, we will utilize the [web5/credentials](https://www.npmjs.com/package/@web5/credentials) package.
```

> 1. Once the customer accepts your Credential Offer, the Wallet will then send a request for credentials to the endpoint specified in the `credential_endpoint` field of the issuer's metadata. Here is an example of the Wallet's request:

Reviewer: ```suggestion
1. Once the customer accepts your Credential Offer, the Wallet will then send a request for credentials to the endpoint specified in the `credential_endpoint` field of the Issuer's metadata. Here is an example of the Wallet's request:
```

```json
POST /credentials HTTP/1.1
Host: issuer.example.com
Authorization: Bearer czZCaGRSa3F0MzpnWDFmQmF0M2JW // access_token
Content-Type: application/json

{
   "proof": {
      "proof_type": "jwt",
      "jwt": "eyJ0eXAiOiJvcGVu...ZoipdP-jvh1WlA"
   }
}

```

The `proof` section contains various claims, for a detailed look at each field please refer to the [OpenID specs](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-7.2.1.1-2.1.1).

> - `czZCaGRSa3F0MzpnWDFmQmF0M2JW`: Is the `access_token` that the Wallet recieved from the `/token` endpoint.

Reviewer: ```suggestion
- `czZCaGRSa3F0MzpnWDFmQmF0M2JW`: The `access_token` that the Wallet recieved from the `/token` endpoint.
```

> :::danger

Reviewer: ```suggestion
:::danger Important
```
